Updated contents and structure (#97)
matter-labs/zksync-web-v2-docs: docs/dev/developer-guides/transactions/blocks.md

2. Timeout has passed. We do not have a precise estimate on the timeout for now.
Users require a timeout season to ensure that their transactions become final on L1 on time.
Imagine if there weren't a timeout, users would send some transactions, and the load would be very low to fill an L1 batch capacity, resulting in users waiting for finality for days.

### Hashes

You may determine the L2 block's hash by knowing the L2 block number because it is deterministic and equivalent to "keccak256(l2_block_number)".
Every block needs a unique hash, which we proposed for Web3 API compatibility. Additionally, to avoid confusing users, we employ the `keccak hash` function to make hashes appear to be random numbers.

### Block Properties
- Timestamp: The current block's creation time as seconds since the Unix epoch
- Block number: The unique sequential number for this block.
- Gas limit: The current block gas limit, always returns `2^32-1`.
- Coinbase:  The current block miner’s address, returns the [bootloader](../contracts/system-contracts.md#bootloader) address.
- Difficulty: The current block difficulty, returns `2500000000000000` (zkSync does not have proof of work consensus).

> - MSIZE: It returns the size of the memory, usually `2^16`.

Reviewer: The `MSIZE` is the number of currently allocated memory for the current contract execution. This is not a block property, but rather a property of a call

# General

> A block in Ethereum is a list of transactions that have been recorded historically. They are collections of transactions that include the hash of the chain's previous block.

Author: What do you suggest?

> A block in Ethereum is a list of transactions that have been recorded historically. They are collections of transactions that include the hash of the chain's previous block.

Reviewer: What do you think about changing:

> A block in Ethereum is a list of transactions that have been recorded historically. They are collections of transactions that include the hash of the chain's previous block.

to

> A block in Ethereum is a list of transactions that have been recorded historically. Besides containing the transactions, the block also includes the hash of the chain's previous block.

## Prerequisite

To fully understand this page, we suggest reading [accounts](account.md), [transactions] and (transactions.md).

## How do Ethereum blocks work?

Blocks are strictly ordered (each new block has a connection to its parent block), and transactions within blocks are strictly ordered to keep track of the history of transactions.
Unless there are exceptional circumstances, all network participants agree with the precise number and history of blocks at any particular time and attempt to combine the pending live transaction requests into the upcoming block.

Every node adds the newly created block to the end of its blockchain after it has been assembled by one of the network's validators, and a new validator is chosen to build the following block.

## Block numbers and time

On zkSync, when a user sends L2 transfers to the operator, they are not processed immediately.
Instead, they are included in a group of pending transactions in the available off-chain state. The user receives witnesses to their balance and all pending transactions in the new rollup state when the operator updates the rollup state. The blockchain does not contain these witnesses.

> The user signs and sends an operation labeled `ProcessTransactions` to the operator to have their pending transactions processed.

Reviewer 2: Maybe I don't fully understand what you mean by witness for the balance, but I think we don't have such functionality. The same for `ProcessTransactions`, `lastSeenBlockNum(sender)`, we do not have it in ZkSync

The operator then includes this operation into the subsequent rollup block, which completes all pending transactions for the sender and changes the on-chain available state to `lastSeenBlockNum(sender) = blockNum`, where `blockNum` is the number of the most recent block. The operator provides witnesses for the updated balances of all users impacted by a rollup block once it has been posted.

### Block Numbers: Ethereum vs. zkSync

The structure of L2 blocks is similar to that of L1 blocks. 

> In zkSync, we send to the L1 contract some information about L2 blocks (a commitment), then we produce a proof for the blocks and transmit it back to the L1 contract, which confirms that the proof is accurate.

Reviewer 2: Instead of "accurate", I'd say "valid" or "correct"

### Ethereum Block Numbers Within zkSync

> Accessing block numbers within a zkSync smart contract (i.e., `block. number` in Solidity) is similar to how you would do it on Ethereum. For example, `eth_blockNumber` returns the number of the latest L2 block, and `eth_getBlockByNumber` returns the information about the L2 block by its number.

Reviewer 2: Here you mixed up block number code property and API methods, that's not correct. Here you'd better write only about API methods.
```suggestion
Accessing block numbers within a zkSync API is similar to how you would do it on Ethereum. For example, `eth_blockNumber` returns the number of the latest L2 block, and `eth_getBlockByNumber` returns the information about the L2 block by its number.
```

> The **capacity,** which includes the number of transactions and storage writes, of the L1 batch is limited because of two different things:

Reviewer 2: You have missed describing what is L1 batch at all and what is the reason why they exist for ZkSync. L1 batch is a range of L2 blocks that are committed and finalized on L1 as one whole. The reason for having them is to save on fixed gas costs while publishing L2 blocks data on the L1 chain. These 2 things definitely should go before you start describing L1 batch details

1. For small L1 batches, a prover can produce proofs in an adequate amount of time.
2. The required gas for the transaction to commit the L1 batch on the L1 contract must not exceed the permitted gas for the L1 block. Because that gas depends on L1 batch size, capacity is constrained.

L1 batches are sealed for two reasons:

> 1. The capacity is reached (well, because we can't send an L1 batch that exceeds capacity).

Author: @perekopskiy what is the capacity that can be reached?

> 1. The capacity is reached (well, because we can't send an L1 batch that exceeds capacity).

Reviewer 2: You can add information that Stat suggests to the paragraph above:
I mean instead of 

1. For small L1 batches, a prover can produce proofs in an adequate amount of time.
2. The required gas for the transaction to commit the L1 batch on the L1 contract must not exceed the permitted gas for the L1 block. Because that gas depends on L1 batch size, capacity is constrained.

write

1. Capacity for zero-knowledge circuits: a prover can produce proofs in an adequate amount of time only for the circuits that fit some conditions.
2. Capacity for the L1 public data: the required gas for the transaction to commit the L1 batch on the L1 contract must not exceed the permitted gas for the L1 block. Because that gas depends on L1 batch size, capacity is constrained.

> 2. Timeout has passed. The timeout is between 1 - 24 hours.

Reviewer 2: I think it is better to mention that we haven't had a more precise estimate yet.

Users require a timeout season to ensure that their transactions become final on L1 at a good time.
The user would send a few transactions if there weren't a timeout, but the load would be pretty low. As a result, it would take a few days to fill an L1 batch capacity, and the user would have to wait for finality throughout these days, which is not great for the user experience.

### Hashes

> You may determine the L2 block's hash by knowing the L2 block number because it is deterministic and equivalent to "keccak256(l2 block number)".

Reviewer 2: ```suggestion
You may determine the L2 block's hash by knowing the L2 block number because it is deterministic and equivalent to "keccak256(l2_block_number)".
```

Every block needs a unique hash, which we mandated for Web3 API compatibility. Additionally, to avoid confusing users, we employ the `keccak hash` function to make hashes appear to be random numbers.

Storage root hash, L1 batch number, and other data are used to determine the hash of the L1 batch. The proofs we supplied to L1 demonstrate that we conduct transactions that cause a state change from the old storage root hash to the committed storage root hash.

> ### Block Properties

Reviewer 2: This paragraph is incorrect, there are many things that we don't have yet on ZkSync, such as block nonce, Beneficiary (because there is only one block producer), mix hash, Transaction root, Receipt root.

I suggest removing it and adding a paragraph `Code block properties`. There you should describe what properties are supported and say that they return properties of L1 batches and not L2 blocks, that's really important. Please ask @StanislavBreadless for the list of the supported properties.

Author: @StanislavBreadless could you help with the list of supported properties?

Reviewer: @krofax you should take a look at the page about the supported/unsupported opcodes. All the block properties on Ethereum that are not listed as unsupported are supported
